patroni/validator.py

```diff
@@ -0,0 +1,325 @@
+
+
+def validate_connect_address(address):
+    if not isinstance(address, str):
+        raise ConfigParseError("is not a string")
+    else:
+        for s in ["127.0.0.1", "0.0.0.0", "*", "::1"]:
+            if s in address:
+                raise ConfigParseError("must not contain {}".format(s))
+    return True
+
+
+def validate_host_port(host_port, listen=False, connect=False):
+    try:
+        host, port = split_host_port(host_port, None)
+    except (ValueError, TypeError):
+        raise ConfigParseError("contains a wrong value")
+    else:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            try:
+                if s.connect_ex((host, port)) == 0 and listen:
+                    ConfigParseError("Port {} is already in use.".format(port))
+                elif connect:
+                    ConfigParseError("{} is not reachable".format(host_port))
+            except socket.gaierror as e:
+                raise ConfigParseError(e)
+    return True
+
+
+def comma_separated_host_port(string):
+    if not isinstance(string, str):
+        raise ConfigParseError("is not a string")
+    assert all([validate_host_port(s.strip()) for s in string.split(",")]), "didn't pass the validation"
+    return True
+
+
+def validate_host_port_listen(host_port):
+    return validate_host_port(host_port, listen=True)
+
+
+def is_ipv4_address(ip):
+    try:
+        a = ip.split(".")
+        assert len(a) == 4
+        assert all([int(i) <= 255 for i in a])
+    except Exception:
+        raise ConfigParseError("{} doesn't look like a valid ipv4 address".format(ip))
+    return True
+
+
+def validate_data_dir(data_dir):
+    if not data_dir:
+        raise ConfigParseError("is an empty string")
+    elif os.path.exists(data_dir) and not os.path.isdir(data_dir):
+        raise ConfigParseError("is not a directory")
+    elif not data_directory_empty(data_dir):
+        if not os.path.exists(os.path.join(data_dir, "PG_VERSION")):
+            raise ConfigParseError("doesn't look like a valid data directory")
+        else:
+            with open(os.path.join(data_dir, "PG_VERSION"), "r") as version:
+                pgversion = int(version.read())
+            waldir = ("pg_wal" if pgversion >= 10 else "pg_xlog")
+            if not os.path.isdir(os.path.join(data_dir, waldir)):
+                raise ConfigParseError("data dir for the cluster is not empty, but doesn't contain"
+                                       " \"{}\" directory".format(waldir))
+    return True
+
+
+def validate_bin_dir(bin_dir):
+    if not bin_dir:
+        raise ConfigParseError("is an empty string")
+    elif not os.path.exists(bin_dir):
+        raise ConfigParseError("Directory '{}' does not exist.".format(bin_dir))
+    elif not os.path.isdir(bin_dir):
+        raise ConfigParseError("'{}' is not a directory.".format(bin_dir))
+    for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup", "postgres"]:
```

Comment:
It would be also great to call the `postgres --version`, parse the output and make sure that it matches with `PG_VERSION`.

The above code parses major version from the output and is able to correctly handle case when the version is in format: `13devel`.
```python
def get_major_version():
    version = subprocess.check_output(['postgres', '--version']).decode()
    version = re.match('^[^\s]+ [^\s]+ (\d+)(\.(\d+))?', version)
    return '.'.join([version.group(1), version.group(3)]) if int(version.group(1)) < 10 else version.group(1)
```